### ISSUE TITLE
Utilization of PETSc solvers

### ISSUE THREAD

**Issue by kenneth-meyer** [OPM/opm-upscaling#321]
Hello,

No real issue here, just a question and looking for advice (and not signed up for mailing list yet). I'm currently exploring replacing some of the DUNE-based solvers in OPM with PETSc based solvers and I came across @jokva 's thesis. To @jokva I just wanted to confirm; this petsc-support branch contains all of the code relevant to your thesis and any further work with utilizing PETSc libraries within opm correct? And what is the best way to clone this/get OPM running on my local system using the petsc library/your code? (apologies if that is a dumb question, I'm still learning and somewhat blindly followed the OPM instruction guide without much understanding of how to change the repo the backend is running from).

Additionally, to all, are there other good examples of where petsc solvers have been leveraged within OPM, and if so do they utilize GPUs (and where could I find these exapmles?). I am very new to the OPM codebase so any help or guidance is greatly appreciated!

Sincerely,
Kenneth Meyer

**Comment by alfbr** [comment]
Hi Kenneth,

It has been a long time since Jørgen's thesis, and both Petsc and GPUs have seen a couple of iterations of implementations since then. Too much to cover in a github comment, so just a short summary from my end. Jørgen's work was limited to upscaling, and showed very interesting results. Bringing this work into Flow, the reservoir simulator, proved to be very challenging. Multiple developers from Norce succeeded however, demonstrating that the Dune solvers are quite competitive (an exercise we have done with multiple solver code bases over the years). GPUs are different beasts. The only successful (meaning actually giving a performance benefit) work there was done by Tong Dong with the help of Nvidia. That implementation is still in the early days. You will find the latest PR here:
https://github.com/OPM/opm-simulators/pull/2628
If you want more details, it will help to know the basis of your interest.

**Comment by jokva** [comment]
> To @jokva I just wanted to confirm; this petsc-support branch contains all of the code relevant to your thesis and any further work with utilizing PETSc libraries within opm correct?

Not really. It did, at the time, provide enough code to be a useful starting point atleast, but a few design choices would probably never have made it upstream, and since then opm has moved a lot.
